{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f7689b4d",
   "metadata": {},
   "source": [
    "# Lab 4\n",
    "\n",
    "- Name:\n",
    "- USC Id:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbecefcf",
   "metadata": {},
   "source": [
    "### 1. Gene expression cancer RNA-Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fd52d0d",
   "metadata": {},
   "source": [
    "Package Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "39fc2b0e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.842587Z",
     "start_time": "2025-11-05T00:17:55.839972Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import OrdinalEncoder\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.naive_bayes import BernoulliNB\n",
    "from sklearn.naive_bayes import CategoricalNB\n",
    "\n",
    "# ===== Optional : import other libraries here ===== #\n",
    "\n",
    "# ===== End of Optional : import other libraries here ===== #"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f15b217",
   "metadata": {},
   "source": [
    "#### Problem (a) Load the files data.csv and labels.csv\n",
    "\n",
    "- They contain the Data Set from: https://archive.ics.uci.edu/dataset/401/gene+expression+cancer+rna+seq\n",
    "\n",
    "- data.csv contains the genetic features for each tumor and labels.csv contains the label of each tumor. After loading the datasets, combine the data sets into a single dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "22f87367",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.857053Z",
     "start_time": "2025-11-05T00:17:55.855077Z"
    }
   },
   "outputs": [],
   "source": [
    "data = None\n",
    "\n",
    "# ===== Read the data.csv file ===== #\n",
    "data = pd.read_csv('data.csv', index_col=0)\n",
    "# ===== End of reading data.csv file ===== #\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f241ed25",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.872429Z",
     "start_time": "2025-11-05T00:17:55.870430Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_0</th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>gene_3</th>\n",
       "      <th>gene_4</th>\n",
       "      <th>gene_5</th>\n",
       "      <th>gene_6</th>\n",
       "      <th>gene_7</th>\n",
       "      <th>gene_8</th>\n",
       "      <th>gene_9</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_20521</th>\n",
       "      <th>gene_20522</th>\n",
       "      <th>gene_20523</th>\n",
       "      <th>gene_20524</th>\n",
       "      <th>gene_20525</th>\n",
       "      <th>gene_20526</th>\n",
       "      <th>gene_20527</th>\n",
       "      <th>gene_20528</th>\n",
       "      <th>gene_20529</th>\n",
       "      <th>gene_20530</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.017209</td>\n",
       "      <td>3.265527</td>\n",
       "      <td>5.478487</td>\n",
       "      <td>10.431999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.175175</td>\n",
       "      <td>0.591871</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.926711</td>\n",
       "      <td>8.210257</td>\n",
       "      <td>9.723516</td>\n",
       "      <td>7.220030</td>\n",
       "      <td>9.119813</td>\n",
       "      <td>12.003135</td>\n",
       "      <td>9.650743</td>\n",
       "      <td>8.921326</td>\n",
       "      <td>5.286759</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.592732</td>\n",
       "      <td>1.588421</td>\n",
       "      <td>7.586157</td>\n",
       "      <td>9.623011</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.816049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.593372</td>\n",
       "      <td>7.323865</td>\n",
       "      <td>9.740931</td>\n",
       "      <td>6.256586</td>\n",
       "      <td>8.381612</td>\n",
       "      <td>12.674552</td>\n",
       "      <td>10.517059</td>\n",
       "      <td>9.397854</td>\n",
       "      <td>2.094168</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.511759</td>\n",
       "      <td>4.327199</td>\n",
       "      <td>6.881787</td>\n",
       "      <td>9.870730</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.972130</td>\n",
       "      <td>0.452595</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.125213</td>\n",
       "      <td>8.127123</td>\n",
       "      <td>10.908640</td>\n",
       "      <td>5.401607</td>\n",
       "      <td>9.911597</td>\n",
       "      <td>9.045255</td>\n",
       "      <td>9.788359</td>\n",
       "      <td>10.090470</td>\n",
       "      <td>1.683023</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.663618</td>\n",
       "      <td>4.507649</td>\n",
       "      <td>6.659068</td>\n",
       "      <td>10.196184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.843375</td>\n",
       "      <td>0.434882</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.076566</td>\n",
       "      <td>8.792959</td>\n",
       "      <td>10.141520</td>\n",
       "      <td>8.942805</td>\n",
       "      <td>9.601208</td>\n",
       "      <td>11.392682</td>\n",
       "      <td>9.694814</td>\n",
       "      <td>9.684365</td>\n",
       "      <td>3.292001</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.655741</td>\n",
       "      <td>2.821547</td>\n",
       "      <td>6.539454</td>\n",
       "      <td>9.738265</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.566967</td>\n",
       "      <td>0.360982</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.996032</td>\n",
       "      <td>8.891425</td>\n",
       "      <td>10.373790</td>\n",
       "      <td>7.181162</td>\n",
       "      <td>9.846910</td>\n",
       "      <td>11.922439</td>\n",
       "      <td>9.217749</td>\n",
       "      <td>9.461191</td>\n",
       "      <td>5.110372</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 20531 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_0    gene_1    gene_2    gene_3     gene_4  gene_5    gene_6  \\\n",
       "sample_0     0.0  2.017209  3.265527  5.478487  10.431999     0.0  7.175175   \n",
       "sample_1     0.0  0.592732  1.588421  7.586157   9.623011     0.0  6.816049   \n",
       "sample_2     0.0  3.511759  4.327199  6.881787   9.870730     0.0  6.972130   \n",
       "sample_3     0.0  3.663618  4.507649  6.659068  10.196184     0.0  7.843375   \n",
       "sample_4     0.0  2.655741  2.821547  6.539454   9.738265     0.0  6.566967   \n",
       "\n",
       "            gene_7  gene_8  gene_9  ...  gene_20521  gene_20522  gene_20523  \\\n",
       "sample_0  0.591871     0.0     0.0  ...    4.926711    8.210257    9.723516   \n",
       "sample_1  0.000000     0.0     0.0  ...    4.593372    7.323865    9.740931   \n",
       "sample_2  0.452595     0.0     0.0  ...    5.125213    8.127123   10.908640   \n",
       "sample_3  0.434882     0.0     0.0  ...    6.076566    8.792959   10.141520   \n",
       "sample_4  0.360982     0.0     0.0  ...    5.996032    8.891425   10.373790   \n",
       "\n",
       "          gene_20524  gene_20525  gene_20526  gene_20527  gene_20528  \\\n",
       "sample_0    7.220030    9.119813   12.003135    9.650743    8.921326   \n",
       "sample_1    6.256586    8.381612   12.674552   10.517059    9.397854   \n",
       "sample_2    5.401607    9.911597    9.045255    9.788359   10.090470   \n",
       "sample_3    8.942805    9.601208   11.392682    9.694814    9.684365   \n",
       "sample_4    7.181162    9.846910   11.922439    9.217749    9.461191   \n",
       "\n",
       "          gene_20529  gene_20530  \n",
       "sample_0    5.286759         0.0  \n",
       "sample_1    2.094168         0.0  \n",
       "sample_2    1.683023         0.0  \n",
       "sample_3    3.292001         0.0  \n",
       "sample_4    5.110372         0.0  \n",
       "\n",
       "[5 rows x 20531 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed76da6a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.881018Z",
     "start_time": "2025-11-05T00:17:55.879292Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "21a4786b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.889673Z",
     "start_time": "2025-11-05T00:17:55.888065Z"
    }
   },
   "outputs": [],
   "source": [
    "labels = None\n",
    "# ===== Read the labels.csv file ===== #\n",
    "labels = pd.read_csv('labels.csv', index_col=0)\n",
    "\n",
    "# ===== End of readsing labels.csv file ===== #\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "00adc414",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.902254Z",
     "start_time": "2025-11-05T00:17:55.900640Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_0</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_1</th>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_2</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_3</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_4</th>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Class\n",
       "sample_0  PRAD\n",
       "sample_1  LUAD\n",
       "sample_2  PRAD\n",
       "sample_3  PRAD\n",
       "sample_4  BRCA"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f69a227f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.911602Z",
     "start_time": "2025-11-05T00:17:55.909973Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Class label counts:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Class\n",
       "BRCA     300\n",
       "KIRC     146\n",
       "LUAD     141\n",
       "PRAD     136\n",
       "COAD      78\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Class label counts:')\n",
    "labels.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "64f8f6fc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.927697Z",
     "start_time": "2025-11-05T00:17:55.925860Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(801, 20531)\n",
      "(801, 1)\n",
      "Test 1 (a) Passed.\n"
     ]
    }
   ],
   "source": [
    "# Test 1 (a) \n",
    "print(data.shape)\n",
    "print(labels.shape)\n",
    "assert(data.shape == (801, 20531))\n",
    "assert(labels.shape == (801,1))\n",
    "\n",
    "print('Test 1 (a) Passed.')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "763a5b34",
   "metadata": {},
   "source": [
    "### Problem 1 (b): Exploratory data analysis\n",
    "- Select the first 640 instances as the training set and the rest of the data as the test set.\n",
    "- Encode the classes as follows BRCA = 0, KIRC = 1, COAD = 2, LUAD = 3, and PRAD = 4. You can use Ordinal Encoder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2bbefdbe",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.944616Z",
     "start_time": "2025-11-05T00:17:55.942718Z"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "training_data, training_class = None, None\n",
    "test_data, test_class = None, None\n",
    "\n",
    "# ===== split data and labels into test and train ===== #\n",
    "training_data = data.iloc[:640]\n",
    "test_data = data.iloc[640:]\n",
    "training_class = labels.iloc[:640]\n",
    "test_class = labels.iloc[640:]\n",
    "\n",
    "\n",
    "# ===== End of split data and labels into test and train ===== #\n",
    "\n",
    "# ===== Use Ordinal Encoder to encode the classes ===== #\n",
    "training_data = OrdinalEncoder().fit_transform(training_data)\n",
    "test_data = OrdinalEncoder().fit_transform(test_data)\n",
    "\n",
    "# ===== End of Use Ordinal Encoder to encode the classes ===== #\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b2e6e7ff",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.956314Z",
     "start_time": "2025-11-05T00:17:55.954693Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  0., 151., 145., ...,  96., 119.,   0.],\n",
       "       [  0., 155., 120., ..., 157.,   8.,   0.],\n",
       "       [  1.,  67.,  50., ..., 104., 145.,   0.],\n",
       "       ...,\n",
       "       [  0.,  77., 111., ...,  69.,  63.,  11.],\n",
       "       [  0.,  53.,  58., ..., 150.,  94.,   0.],\n",
       "       [  0.,  38., 117., ..., 117.,  60.,   0.]])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baa608fd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.964759Z",
     "start_time": "2025-11-05T00:17:55.962982Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "775a5a20",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.973594Z",
     "start_time": "2025-11-05T00:17:55.971705Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(640, 20531)\n",
      "(161, 20531)\n",
      "(640, 1)\n",
      "(161, 1)\n",
      "Test 1 (b) Passed\n"
     ]
    }
   ],
   "source": [
    "# Test 1 (b)\n",
    "print(training_data.shape)\n",
    "print(test_data.shape)\n",
    "print(training_class.shape)\n",
    "print(test_class.shape)\n",
    "assert(training_data.shape[0] == 640 and training_data.shape[1] == 20531)\n",
    "assert(test_data.shape[0] == 161 and test_data.shape[1] == 20531)\n",
    "assert(training_class.shape[0] == 640)\n",
    "assert(test_class.shape[0] == 161)\n",
    "\n",
    "print(f\"Test 1 (b) Passed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "0c7329a8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:55.993074Z",
     "start_time": "2025-11-05T00:17:55.991466Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_0</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_1</th>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_2</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_3</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_4</th>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_635</th>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_636</th>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_637</th>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_638</th>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_639</th>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>640 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Class\n",
       "sample_0    PRAD\n",
       "sample_1    LUAD\n",
       "sample_2    PRAD\n",
       "sample_3    PRAD\n",
       "sample_4    BRCA\n",
       "...          ...\n",
       "sample_635  LUAD\n",
       "sample_636  LUAD\n",
       "sample_637  BRCA\n",
       "sample_638  BRCA\n",
       "sample_639  BRCA\n",
       "\n",
       "[640 rows x 1 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "training_class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "08562817",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.001960Z",
     "start_time": "2025-11-05T00:17:56.000189Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_640</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_641</th>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_642</th>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_643</th>\n",
       "      <td>KIRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_644</th>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_796</th>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_797</th>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_798</th>\n",
       "      <td>COAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_799</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_800</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Class\n",
       "sample_640  PRAD\n",
       "sample_641  BRCA\n",
       "sample_642  BRCA\n",
       "sample_643  KIRC\n",
       "sample_644  LUAD\n",
       "...          ...\n",
       "sample_796  BRCA\n",
       "sample_797  LUAD\n",
       "sample_798  COAD\n",
       "sample_799  PRAD\n",
       "sample_800  PRAD\n",
       "\n",
       "[161 rows x 1 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_class"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a021c7ef",
   "metadata": {},
   "source": [
    "### Problem 1 (c) Classification using Gaussian Naive Bayes\n",
    "\n",
    "- Use sklearn’s Gaussian Naive Bayes method to build a classifier based on training data. Report the training misclassification error rate (the percentage of training data that are misclassified)\n",
    "\n",
    "- Use sklearn’s Gaussian Naive Bayes method to classify test data, using the model you developed in 1(c). Report test misclassification error rate (the percentage of training data that are misclassified)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "248c9332",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.010052Z",
     "start_time": "2025-11-05T00:17:56.008448Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/anaconda3/envs/py312/lib/python3.12/site-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  y = column_or_1d(y, warn=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> Gaussian NB\n",
      "Training Misclassification Rate: 0.0015625\n"
     ]
    }
   ],
   "source": [
    "# ===== Train Gaussian NB model ===== #\n",
    "model = GaussianNB()\n",
    "model.fit(training_data, training_class)\n",
    "\n",
    "# ===== End of Train Gaussian NB model ===== #\n",
    "\n",
    "\n",
    "# ===== Use the model to predict labels for the training data ===== #\n",
    "training_predictions = model.predict(training_data)\n",
    "training_misclassification_rate = np.mean(training_predictions != training_class.values.ravel())\n",
    "# ===== End of Use the model to predict labels for the training data ===== #\n",
    "print(f\">>> Gaussian NB\")\n",
    "print(f\"Training Misclassification Rate: {training_misclassification_rate}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "96aada64",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.023658Z",
     "start_time": "2025-11-05T00:17:56.022012Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> Gaussian NB\n",
      "Test Misclassification Rate: 0.2795031055900621\n"
     ]
    }
   ],
   "source": [
    "test_misclassification_rate = None\n",
    "\n",
    "# ===== Use the model to predict labels for the test data ===== #\n",
    "test_predictions = model.predict(test_data)\n",
    "test_misclassification_rate = np.mean(test_predictions != test_class.values.ravel())\n",
    "\n",
    "# ===== End of Use the model to predict labels for the test data ===== #\n",
    "\n",
    "print(f\">>> Gaussian NB\")\n",
    "print(f\"Test Misclassification Rate: {test_misclassification_rate}\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce9fbd87",
   "metadata": {},
   "source": [
    "### Ploblem 1 (d): Classification using Bernoulli Naive Bayes\n",
    "\n",
    "- Calculate the median of each of the gene features. Binarize the features: any feature greater or equal than the median of that feature must be converted to 1 and any feature less than or equal to the median must be converted to zero. (5 pts)\n",
    "\n",
    "- Use sklearn’s Bernoulli Na ̈ıve Bayes method with Laplace Smoothing To build a classifier based on binarized training data. Report the training misclassification error rate (the percentage of training data that are misclassified). (20 pts)\n",
    "\n",
    "- Use sklearn’s Bernoulli Na ̈ıve Bayes method to classify test data, using the model you developed in 1(d)ii. Report test misclassification error rate (the percentage of training data that are misclassified). (20 pts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "a1ce71a1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.033502Z",
     "start_time": "2025-11-05T00:17:56.032119Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1.00000000e-10 3.14368728e+00 3.12700588e+00 ... 9.56651076e+00\n",
      " 5.21861845e+00 1.00000000e-10]\n"
     ]
    }
   ],
   "source": [
    "binarized_data = None\n",
    "binarized_training_data = None\n",
    "binarized_test_data = None\n",
    "\n",
    "# ===== Calculate the medians for each feature in the trainind data and binarize the values for each column  ===== #\n",
    "\n",
    "medians = np.median(data, axis=0) + 1e-10 #avoid median = 0, that all columns will be 1\n",
    "print(medians)\n",
    "binarized_training_data = (training_data >= medians).astype(int)\n",
    "binarized_test_data = (test_data >= medians).astype(int)\n",
    "\n",
    "# Convert to DataFrame for easier handling (optional, but matches the expected format)\n",
    "binarized_training_data = pd.DataFrame(binarized_training_data, \n",
    "                                     columns=[f'gene_{i+1}' for i in range(training_data.shape[1])])\n",
    "binarized_test_data = pd.DataFrame(binarized_test_data, \n",
    "                                 columns=[f'gene_{i+1}' for i in range(test_data.shape[1])])\n",
    "\n",
    "# ===== End of Calculate the medians for each feature and binarize the values for each column  ===== #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "1273f09f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.041225Z",
     "start_time": "2025-11-05T00:17:56.039657Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>gene_3</th>\n",
       "      <th>gene_4</th>\n",
       "      <th>gene_5</th>\n",
       "      <th>gene_6</th>\n",
       "      <th>gene_7</th>\n",
       "      <th>gene_8</th>\n",
       "      <th>gene_9</th>\n",
       "      <th>gene_10</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_20522</th>\n",
       "      <th>gene_20523</th>\n",
       "      <th>gene_20524</th>\n",
       "      <th>gene_20525</th>\n",
       "      <th>gene_20526</th>\n",
       "      <th>gene_20527</th>\n",
       "      <th>gene_20528</th>\n",
       "      <th>gene_20529</th>\n",
       "      <th>gene_20530</th>\n",
       "      <th>gene_20531</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 20531 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_1  gene_2  gene_3  gene_4  gene_5  gene_6  gene_7  gene_8  gene_9  \\\n",
       "0       0       1       1       1       1       0       1       1       0   \n",
       "1       0       0       1       1       1       0       1       0       0   \n",
       "2       0       1       1       1       1       0       1       1       0   \n",
       "3       0       1       1       1       1       0       1       1       0   \n",
       "4       0       1       1       1       1       0       1       1       0   \n",
       "\n",
       "   gene_10  ...  gene_20522  gene_20523  gene_20524  gene_20525  gene_20526  \\\n",
       "0        0  ...           1           1           1           1           1   \n",
       "1        0  ...           1           0           1           1           0   \n",
       "2        0  ...           1           1           1           1           1   \n",
       "3        0  ...           1           1           1           1           1   \n",
       "4        0  ...           1           1           1           1           1   \n",
       "\n",
       "   gene_20527  gene_20528  gene_20529  gene_20530  gene_20531  \n",
       "0           1           1           1           1           0  \n",
       "1           1           1           1           1           0  \n",
       "2           0           1           1           1           0  \n",
       "3           1           1           1           1           0  \n",
       "4           1           1           1           1           0  \n",
       "\n",
       "[5 rows x 20531 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binarized_training_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "bbe351ff",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.051992Z",
     "start_time": "2025-11-05T00:17:56.050131Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>gene_3</th>\n",
       "      <th>gene_4</th>\n",
       "      <th>gene_5</th>\n",
       "      <th>gene_6</th>\n",
       "      <th>gene_7</th>\n",
       "      <th>gene_8</th>\n",
       "      <th>gene_9</th>\n",
       "      <th>gene_10</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_20522</th>\n",
       "      <th>gene_20523</th>\n",
       "      <th>gene_20524</th>\n",
       "      <th>gene_20525</th>\n",
       "      <th>gene_20526</th>\n",
       "      <th>gene_20527</th>\n",
       "      <th>gene_20528</th>\n",
       "      <th>gene_20529</th>\n",
       "      <th>gene_20530</th>\n",
       "      <th>gene_20531</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161 rows × 20531 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_1  gene_2  gene_3  gene_4  gene_5  gene_6  gene_7  gene_8  gene_9  \\\n",
       "0         0       1       1       1       1       0       1       1       0   \n",
       "1         0       1       1       1       1       0       1       1       0   \n",
       "2         1       1       1       1       1       0       1       1       0   \n",
       "3         0       1       1       1       0       0       1       1       0   \n",
       "4         0       1       1       1       1       0       1       0       0   \n",
       "..      ...     ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "156       0       1       1       1       1       0       1       1       0   \n",
       "157       0       1       1       1       1       0       1       0       0   \n",
       "158       0       1       1       1       1       0       1       1       0   \n",
       "159       0       1       1       1       1       0       1       0       0   \n",
       "160       0       1       1       1       1       0       1       0       0   \n",
       "\n",
       "     gene_10  ...  gene_20522  gene_20523  gene_20524  gene_20525  gene_20526  \\\n",
       "0          0  ...           1           1           1           1           1   \n",
       "1          0  ...           1           1           1           1           1   \n",
       "2          0  ...           1           1           1           1           1   \n",
       "3          0  ...           1           1           1           1           1   \n",
       "4          0  ...           1           0           1           1           1   \n",
       "..       ...  ...         ...         ...         ...         ...         ...   \n",
       "156        0  ...           1           1           1           1           1   \n",
       "157        0  ...           1           1           1           1           1   \n",
       "158        0  ...           1           1           1           1           1   \n",
       "159        0  ...           1           1           1           1           1   \n",
       "160        0  ...           1           1           1           1           1   \n",
       "\n",
       "     gene_20527  gene_20528  gene_20529  gene_20530  gene_20531  \n",
       "0             1           1           1           1           0  \n",
       "1             1           1           1           1           0  \n",
       "2             0           1           1           1           0  \n",
       "3             1           1           1           1           0  \n",
       "4             1           1           0           1           1  \n",
       "..          ...         ...         ...         ...         ...  \n",
       "156           1           1           1           1           0  \n",
       "157           1           1           1           1           0  \n",
       "158           1           1           1           1           1  \n",
       "159           1           1           1           1           0  \n",
       "160           1           1           1           1           0  \n",
       "\n",
       "[161 rows x 20531 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binarized_test_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "6dd0ea7b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.088771Z",
     "start_time": "2025-11-05T00:17:56.087202Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "Test Passed\n"
     ]
    }
   ],
   "source": [
    "print(len(binarized_training_data['gene_1'].value_counts()))\n",
    "assert (binarized_training_data.shape == (640,20531))\n",
    "assert (binarized_test_data.shape == (161,20531))\n",
    "assert (len(binarized_training_data['gene_1'].value_counts()) == 2)\n",
    "print('Test Passed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76bf2e9f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.118224Z",
     "start_time": "2025-11-05T00:17:56.116718Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "ffc5f86c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.141304Z",
     "start_time": "2025-11-05T00:17:56.139505Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> Bernoulli NB\n",
      "Training Misclassification Rate: 0.0015625\n"
     ]
    }
   ],
   "source": [
    "# ===== Train Bernoulli NB model ===== #\n",
    "# Train Bernoulli Naïve Bayes classifier with Laplace smoothing (alpha=1 by default)\n",
    "model = BernoulliNB(alpha=1.0)\n",
    "model.fit(binarized_training_data, training_class.values.ravel() )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# ===== End of Train Bernoulli NB model ===== #\n",
    "\n",
    "training_misclassification_rate = None\n",
    "\n",
    "# ===== Use the model to predict labels for the training data ===== #\n",
    "training_preds = model.predict(binarized_training_data)\n",
    "training_misclassification_rate = np.mean(training_preds != training_class.values.ravel())\n",
    "\n",
    "\n",
    "# ===== End of Use the model to predict labels for the training data ===== #\n",
    "\n",
    "print(f\">>> Bernoulli NB\")\n",
    "print(f\"Training Misclassification Rate: {training_misclassification_rate}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "a8a57849",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.169061Z",
     "start_time": "2025-11-05T00:17:56.167530Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> Bernoulli NB\n",
      "Test Misclassification Rate: 0.024844720496894408\n"
     ]
    }
   ],
   "source": [
    "test_misclassification_rate = None\n",
    "\n",
    "# ===== Use the model to predict labels for the test data ===== #\n",
    "test_preds = model.predict(binarized_test_data)\n",
    "test_misclassification_rate = np.mean(test_preds != test_class.values.ravel())\n",
    "\n",
    "# ===== End of Use the model to predict labels for the test data ===== #\n",
    "\n",
    "print(f\">>> Bernoulli NB\")\n",
    "print(f\"Test Misclassification Rate: {test_misclassification_rate}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b26d645f",
   "metadata": {},
   "source": [
    "### 2. Extra Credit: Categorical Naive Bayes\n",
    "\n",
    "Create 10 equally spaced bins between the maximum and minimum of each feature in the training set and convert the features to categorical training data using those bins. Convert the test data based on the bins you calculated for training data into categorical features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3979c30c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.194240Z",
     "start_time": "2025-11-05T00:17:56.192223Z"
    }
   },
   "outputs": [],
   "source": [
    "# binned_training_data = pd.DataFrame()\n",
    "# binned_test_data = pd.DataFrame()\n",
    "\n",
    "# Hint: Use np.linpsace to create bin boundaries and pd.cut to segregate data into the bins\n",
    "# ===== For each feature create 10 bins using the training data and convert test and train data into categorical values  ===== #\n",
    "\n",
    "\n",
    "    \n",
    "# ===== End of For each feature create 10 bins using the training data and convert test and train data into categorical values  ===== #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ac0c0cc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.217012Z",
     "start_time": "2025-11-05T00:17:56.215655Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05d7f459",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.241811Z",
     "start_time": "2025-11-05T00:17:56.240242Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a5173fa6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.263533Z",
     "start_time": "2025-11-05T00:17:56.261962Z"
    }
   },
   "outputs": [],
   "source": [
    "# ===== Train Categorical NB model ===== #\n",
    "\n",
    "\n",
    "\n",
    "# ===== End of Train Categorical NB model ===== #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8da92ca4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.296394Z",
     "start_time": "2025-11-05T00:17:56.294350Z"
    }
   },
   "outputs": [],
   "source": [
    "# training_misclassification_rate = None\n",
    "\n",
    "# ===== Use the model to predict labels for the training data ===== #\n",
    "\n",
    "\n",
    "\n",
    "# ===== End of Use the model to predict labels for the training data ===== #\n",
    "\n",
    "# print(f\">>> Categorical NB\")\n",
    "# print(f\"Training Misclassification Rate: {training_misclassification_rate}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "32a3a9d6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.321676Z",
     "start_time": "2025-11-05T00:17:56.320121Z"
    }
   },
   "outputs": [],
   "source": [
    "# test_misclassification_rate = None\n",
    "\n",
    "# ===== Use the model to predict labels for the test data ===== #\n",
    "\n",
    "\n",
    "# ===== End of Use the model to predict labels for the test data ===== #\n",
    "\n",
    "# print(f\">>> Categorical NB\")\n",
    "# print(f\"Test Misclassification Rate: {test_misclassification_rate}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f92db404",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-05T00:17:56.345973Z",
     "start_time": "2025-11-05T00:17:56.344733Z"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
